tutorials/workshop/Workshop_2_Writing_Labeling_Functions.ipynb

```diff
@@ -0,0 +1,502 @@
+    "###  Using a _Development Set_ of Human-labeled Data\n",
+    "\n",
+    "In our setting, we will use the phrase _development set_ to refer to a set of examples (here, a subset of our training set) which we label by hand and use to help us develop and refine labeling functions.  Unlike the _test set_, which we do not look at and use for final evaluation, we can inspect the development set while writing labeling functions. This is a list of `{-1,1}` labels."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pickle\n",
+    "\n",
+    "with open('new_dev_data.pkl', 'rb') as f:\n",
+    "    dev_data = pickle.load(f)\n",
+    "    dev_labels = pickle.load(f)\n",
+    "    \n",
+    "with open('new_train_data.pkl', 'rb') as f:\n",
+    "    train_data = pickle.load(f)\n",
+    "    train_labels = pickle.load(f)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Labeling Function Metrics\n",
+    "\n",
+    "#### Coverage\n",
+    "One simple metric we can compute quickly is our _coverage_, the number of candidates labeled by our LF, on our training set (or any other set).\n",
+    "\n",
+    "#### Precision / Recall / F1\n",
+    "If we have gold labeled data, we can also compute standard precision, recall, and F1 metrics for the output of a single labeling function. These metrics are computed over 4 _error buckets_: _True Positives_ (tp), _False Positives_ (fp), _True Negatives_ (tn), and _False Negatives_ (fn).\n",
+    "\n",
+    "\\begin{equation*}\n",
+    "precision = \\frac{tp}{(tp + fp)}\n",
+    "\\end{equation*}\n",
+    "\n",
+    "\\begin{equation*}\n",
+    "recall = \\frac{tp}{(tp + fn)}\n",
+    "\\end{equation*}\n",
+    "\n",
+    "\\begin{equation*}\n",
+    "F1 = 2 \\cdot \\frac{ (precision \\cdot recall)}{(precision + recall)}\n",
+    "\\end{equation*}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# II. Labeling Functions\n",
+    "\n",
+    "## A. Pattern Matching Labeling Functions\n",
+    "\n",
+    "One powerful form of labeling function design is defining sets of keywords or regular expressions that, as a human labeler, you know are correlated with the true label. For example, we could define a dictionary of terms that occur between person names in a candidate. One simple dictionary of terms indicating a true relation could be:\n",
+    "    \n",
+    "    spouses = {'husband', 'wife'}\n",
+    " \n",
+    "We can then write a labeling function that checks for a match with these terms in the text that occurs between person names. To access the text between the person mentions, we can use the `get_between_tokens` preprocessors from the previous notebook.\n",
+    "\n",
+    "    @labeling_function(resources=dict(spouses=spouses), preprocessors=[get_between_tokens])\n",
+    "    def LF_husband_wife(x: DataPoint, spouses: List[str]) -> int:\n",
```

Comment:
Let's avoid typing in the tutorials

Reply:
These are also incorrectly typed (should be a Set)

Reply:
would this be
`def LF_husband_wife(x, spouses) -> int` or without the ` -> int` part too?